grpc/cubic.proto

```diff
@@ -175,119 +148,135 @@ message RecognitionConfig {
   // Unique identifier of the model to use, as obtained from a `Model` message.
   string model_id = 1;
 
-  // The encoding of the audio data to be sent for recognition.
+  // Format of audio data sent/streamed through the `RecognitionAudio` messages.
+  // For formats like WAV/MP3 that have headers, the headers are expected to be
+  // sent at the beginning of the stream, not in every `RecognitionAudio`
+  // message.
+  //
+  // An error is returned if the audio format is not specified.
   //
-  // For best results, the audio source should be captured and transmitted using
-  // the RAW_LINEAR16 encoding.
-  enum Encoding {
-    // Raw (headerless) Uncompressed 16-bit signed little endian samples (linear
-    // PCM), single channel, sampled at the rate expected by the chosen `Model`.
-    RAW_LINEAR16 = 0;
+  // Depending on how they are configured, server instances of this service may
+  // not support all the formats provided in this API.
+  //
+  // For best results, use the RAW audio with Little-Endian SignedInteger PCM
+  // encoding of audio sampled at the same rate as the model you are using.
+  oneof audio_format {
+    AudioFormatRAW audio_format_raw = 2;
+    AudioFormatWithHeaders audio_format_with_headers = 3;
+  }
 
-    // WAV (data with RIFF headers), with data sampled at a rate equal to or
-    // higher than the sample rate expected by the chosen Model.
-    WAV = 1;
+  message AudioFormatRAW {
+    // Sample Rate of the audio in Hertz (e.g. 16000, 8000)
+    uint32 sample_rate = 1;
 
-    // MP3 data, sampled at a rate equal to or higher than the sampling rate
-    // expected by the chosen Model.
-    MP3 = 2;
+    // Bit Depth of each audio sample (e.g. 8, 16, etc.) This field is ignored
+    // for fixed-length encoding such as ULAW.
+    uint32 bit_depth = 2;
 
-    // FLAC data, sampled at a rate equal to or higher than the sample rate
-    // expected by the chosen Model.
-    FLAC = 3;
+    // Encoding of the audio.
+    Encoding encoding = 3;
+
+    // When audio samples have a bit-depth > 8, the data is assumed to be
+    // little-endian, unless this field is set to true.
+    bool big_endian = 4;
 
-    // VOX data (Dialogic ADPCM), sampled at 8 KHz.
-    VOX8000 = 4;
+    // Number of channels present in the audio.
+    uint32 num_channels = 5;
```

Comment:
Do we need to worry about channels being interleaved in the raw data, and specifying somehow the way they are interleaved?

Reply:
I think not. I've only seen one way of laying channels out in the stream.